### PR TITLE
Make Query::new() pub

### PR DESCRIPTION
**Pull request by JMS55**
## Changelog
- Added `Query::new()` for low-level, unsafe uses of the ECS world.

### CODE REVIEW

**Comment by Nilirad** [comment]
Should this be controversial? It makes a private API public. But it is already marked as `unsafe`, so is that enough?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I think we should add docs to the effect of "generally speaking, if you need a `Query` from ` World`, you should use `SystemState` instead."

But this is a reasonable API for `UnsafeWorldCell`, and I expect advanced users will be alright with breakage.

File: crates/bevy_ecs/src/system/query.rs [hunk 1]
```diff
@@ -363,7 +363,7 @@ impl<'w, 's, Q: WorldQueryData, F: WorldQueryFilter> Query<'w, 's, Q, F> {
     /// This will create a query that could violate memory safety rules. Make sure that this is only
     /// called in ways that ensure the queries have unique mutable access.
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/system/query.rs:364]:
```suggestion
    /// `world` must have permission to access any component data specified in `state`'s [`archetype_component_access`].
    ///
    /// [`archetype_component_access`]: QueryState::archetype_component_access
    ///
    /// # Panics
    ///
    /// If `world` does not match the one used to create `state`.
```

I believe this is the most correct way you can phrase the safety requirement. Though we'd also need to add `QueryState::archetype_component_access` (it's weird that function doesn't exist already).

The `# Panics` should probably go above the `# Safety` section, although GitHub's UI won't let me format the suggestion that way.

**Review by JoJoJet — COMMENTED** [review]
I *really* don't want to include the `force_read_only_component_access` hack in a public API. Can we rename the existing method to `new_inner`, and make `pub fn new` a thin wrapper around `new_inner`?

**Review by james7132 — COMMENTED** [review]
I'll be the one to play devil's advocate and ask what the expected use case here is. `SystemState::get` and it's variants should be sufficient, and provides a list of options that can be used to create a `Query` with a gradual introduction of increasingly tighter invariants, and it ensures that `QueryState` stays in a proper state while in use. Is there any particular use case that requires `Query::new` to be public that isn't served by this flow?

**Comment by JMS55** [comment]
For reference, here's what I'm trying to achieve: https://github.com/JMS55/bevy_dioxus/blob/main/src/hooks.rs#L104-L125
```rust
fn ui_root(cx: Scope) -> Element {
    let names = use_query::<DebugName>(cx);

    render!(
        for name in names.query() {
            "{name:?}"
        }
    )
}
```

**Comment by JMS55** [comment]
Using SystemState worked, but has the unfortunate side effect of requiring things to be `mut` https://github.com/JMS55/bevy_dioxus/pull/19

**Pull request closed without merging**